rfd/0163-vnet.md

```diff
@@ -0,0 +1,996 @@
+```
+
+```proto
+// VNetService provides an API to manage VNets.
+service VNetService {
+  // GetVNet returns the specified VNet resource.
+  rpc GetVNet(GetVNetRequest) returns (VNet);
+
+  // CreateVNet creates a new VNet resource.
+  rpc CreateVNet(CreateVNetRequest) returns (VNet);
+
+  // UpdateVNet updates an existing VNet resource.
+  rpc UpdateVNet(UpdateVNetRequest) returns (VNet);
+
+  // DeleteVNet hard deletes the specified VNet resource.
+  rpc DeleteVNet(DeleteVNetRequest) returns (google.protobuf.Empty);
+}
+
+// Request for GetVNet.
+message GetVNetRequest {}
+
+// Request for CreateVNet.
+message CreateVNetRequest {
+  // The VNet resource to create.
+  VNet vnet = 1;
+}
+
+// Request for UpdateVNet.
+message UpdateVNetRequest {
+  // The VNet resource to create.
+  VNet vnet = 1;
+
+  // The update mask applied to a VNet.
+  // Fields are masked according to their proto name.
+  FieldMask update_mask = 2;
+}
+
+// Request for DeleteVNet.
+message DeleteVNetRequest {}
+```
+
+#### Daemon service
+
+```proto
+service VNetRootDaemon {
+  // Start creates and configures a TUN device and configures DNS for all active profiles found in
+  // tsh_home. It also saves pid and uid of the client. Until the client calls Stop or exits, no
+  // other client can call Start or Stop.
+  //
+  // Start returns the fd of the TUN device so that the client can read and write from it. This is
+  // done through the underlying socket connection, see https://pkg.go.dev/github.com/edwarnicke/grpcfd.
+  rpc Start(StartRequest) returns (StartResponse);
+
+  // RefreshDNSConfig re-fetches cluster information and updates the existing DNS configuration to
+  // account for any changes. Typically called by the client after it detects that the user has
+  // logged in or out of a cluster.
+  rpc RefreshDNSConfig(RefreshDNSConfigRequest) returns (RefreshDNSConfigResponse);
+
+  // Stop closes any active connections, closes the TUN device and cleans up DNS configuration.
+  // This is also done automatically if the daemon detects that the client has exited without
+  // explicitly calling Stop.
+  rpc Stop(StopRequest) returns (StopResponse);
+}
+
+// Empty messages such as StopRequest have been omitted.
+
+message StartRequest {
+  // tsh_home is a path that points to the tsh home directory. VNet creates proxy clients for active
+  // profiles in this directory and configures DNS for each cluster found.
+  //
+  // tsh_home must belong to the client, as verified by uid of net.Conn.
+  string tsh_home = 1;
+}
+```
+
+### Binary size
+
+VNet adds new dependencies on `golang.zx2c4.com/wireguard` and
+`gvisor.dev/gvisor`.
+The demo implementation currently increases the size of local builds of the
+`tsh` binary from 106M to 109M.
+
+### Backward Compatibility
+
+We will not be changing or interfering with existing app access to backward
+compatibility is not an issue.
+
+### Audit Events
+
+VNet is a client-side feature so no audit events will be added.
```

Comment:
Perhaps I had a misunderstanding.  Are IP assignments only specific to the individual client?  If so, why call out the potential for direct IP usage at all?

In any case a couple cluster audit events come to mind:
* IP assignments (if shared)
* Custom DNS Zone configuration changes

On the client side we should ensure that daemon also has a complete record of activity which is protected from the user.

Reply:
Yeah I wrote this before adding the `vnet` cluster configuration resource, Create/Update/Delete of that will get audit events

> On the client side we should ensure that daemon also has a complete record of activity which is protected from the user

Can you expand on this? If it is protected from the user who can see it?

Reply:
Perhaps the daemon is not the right place to capture this logging, but my thought is "what would be required to investigate a network level attack through vnet".  My concern is that the vnet IP range could potentially be used to hide the network destinations requested to.

If IP addresses are globaly managed, this seems particularly important to understand at the cluster administration level.  If they are local, then I think we should leverage the privileged local daemon to have a record of dns activity on the vnet.

What do you think is the best way to ensure vnet does not introduce an obfuscation into what network activity is occuring?

Reply:
The vnet IPs _never leave the user's workstation_. It's a purely virtual client-side concept. The network traffic on the wire is exactly the same as it is today, so I don't see any concerns about obfuscation.

You'd investigate a network-level attack through vnet the same way you'd investigate a network-level attack with `tsh proxy app` - by looking at the client IPs in the audit events. (The true IPs, as seen by the proxy, not the vnet ones).